input-output-hk/daedalus
# [DDW-245] Download manager

This PR implements a download manager.

## Todos

- [x] Method for requesting a new download
- [x] Method for requesting the local data for a stopped download
- [x] Method for resuming a stopped download

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1591303142025200?thread_ts=1591303119.024900&cid=GGKFXSKC6)
* Steps for testing:
  * Open the Devtools, Console tab;
  * Open the State directory (it can be found in the diagnostics screen);
  * Run the command `daedalus.stores.appUpdate._requestUpdateDownload()`;
  * Inside the `Downloads` folder a file with a name format of `Unconfirmed-{timestamp}.crdownload` should appear;
  * The download progress should be printed in the Console;
  * Close the app when the download is about 50%;
  * Reopen the app;
  * Run the command `daedalus.stores.appUpdate._managePendingUpdate()` in the Devtools console;
  * A `downloadLocalData` object should be printed, containing the download's data and progress information;
  * Run the command `daedalus.stores.appUpdate._requestResumeUpdateDownload()`;
  * The download should resume from the correct percentage progress;
  * After the download is finished, the file inside the Downloads folder should have been renamed to `daedalus-1.1.0-mainnet-12849.pkg`.

* Obs: this task doesn't handle the app update itself. It's a generic Download Manager.


---

## Review Checklist

### Basics

- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese

### Code Quality

### Testing

### After Review
